java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -175,13 +193,74 @@ class Guard extends ExprParent {
   }
 }
 
+/**
+ * A `Guard` that tests an expression's type -- that is, an `instanceof T` or a
+ * `case T varname` pattern case.
+ */
+class TypeTestGuard extends Guard {
+  Expr testedExpr;
+  Type testedType;
+
+  TypeTestGuard() {
+    exists(InstanceOfExpr ioe | this = ioe |
+      testedExpr = ioe.getExpr() and
+      testedType = ioe.getSyntacticCheckedType()
+    )
+    or
+    exists(PatternCase pc | this = pc |
+      pc.getSelectorExpr() = testedExpr and
+      testedType = pc.getPattern().getType()
+    )
+  }
+
+  /**
+   * Gets the record pattern this type test binds to, if any.
+   */
+  PatternExpr getPattern() {
+    result = this.(InstanceOfExpr).getPattern()
+    or
+    result = this.(PatternCase).getPattern()
+  }
+
+  /**
+   * Holds if this guard tests whether `e` has type `t` on `testedBranch`.
+   *
+   * Note that record patterns that make at least one tighter restriction than the record's definition
+   * (e.g. matching `record R(Object)` with `case R(String)`) means this only guarantees the tested type
+   * on the true branch (i.e., entering such a case guarantees `testedExpr` is a `testedType`, but failing
+   * the type test could mean a nested record or binding pattern didn't match but `testedExpr` is still
+   * of type `testedType`.)
+   */
+  predicate appliesTypeTest(Expr e, Type t, boolean testedBranch) {
```

Comment:
The `testedBranch` column in this predicate is always either `[true,false]` or just `true`. Maybe we should change its name and represent those two possibilities as `true` and `false`, respectively. That way we don't duplicate all the `(Guard, Expr, Type)` tuples for which the test is an unrestricted pattern. Maybe a good name would be something like "faithful"?